uc-crux-llvm/src/UCCrux/LLVM/View/Cursor.hs

```diff
@@ -0,0 +1,127 @@
+{-
+Module           : UCCrux.LLVM.View.Cursor
+Description      : See "UCCrux.LLVM.View".
+Copyright        : (c) Galois, Inc 2022
+License          : BSD3
+Maintainer       : Langston Barrett <langston@galois.com>
+Stability        : provisional
+-}
+
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE StrictData #-}
+{-# LANGUAGE TemplateHaskell #-}
+
+module UCCrux.LLVM.View.Cursor
+  ( ViewCursorError,
+    ppViewCursorError,
+    CursorView(..),
+    cursorView,
+    viewCursor,
+  )
+where
+
+import           Control.Lens ((^.))
+import           Control.Monad (when)
+import qualified Data.Aeson as Aeson
+import qualified Data.Aeson.TH as Aeson.TH
+import           GHC.Generics (Generic)
+import           Numeric.Natural (Natural)
+
+import           Prettyprinter (Doc)
+import qualified Prettyprinter as PP
+
+import           Data.Parameterized.Classes (IxedF'(ixF'))
+import qualified Data.Parameterized.Context as Ctx
+import qualified Data.Parameterized.NatRepr as NatRepr
+import           Data.Parameterized.Some (Some(Some))
+
+import           UCCrux.LLVM.Cursor (Cursor(..))
+import           UCCrux.LLVM.FullType.Type (FullTypeRepr(..), ModuleTypes, asFullType)
+
+data ViewCursorError
+  = StructBadIndex !Int
+  | TypeMismatch
+  | VectorLengthMismatch !Natural !Natural
+  | VectorBadIndex !Natural !Natural
+  deriving (Eq, Ord, Show)
+
+ppViewCursorError :: ViewCursorError -> Doc ann
+ppViewCursorError =
+  \case
+    StructBadIndex i -> "Bad struct index: " <> PP.viaShow i
+    TypeMismatch -> "Type mismatch"
+    VectorLengthMismatch typeLen len ->
+      PP.hsep
+        [ "Vector length mismatch. Expected length",
+          PP.viaShow typeLen,
+          "but got length",
+          PP.viaShow len
+        ]
+    VectorBadIndex idx len ->
+      PP.hsep
+        [ "Bad vector index. Found index",
+          PP.viaShow idx,
+          "but vector was of length",
+          PP.viaShow len
+        ]
+
+data CursorView
+  = VHere
+  | VDereference Int CursorView
+  | VIndex Natural Natural CursorView
+  | VField Int CursorView
```

Comment:
This set of views is prefixed with `V`, while some of the others are suffixed with `View`. Is there a logic to the name choices?

Reply:
Apparently I decided to use a prefix for sum types and a suffix for record types. Fixed to be consistent!